pydicom/dataset.py

```diff
@@ -355,78 +362,162 @@ def __iter__(self):
+            gdcm_image_reader.SetFileName(self.filename)
+            if not gdcm_image_reader.Read():
+                raise TypeError("GDCM could not read DICOM image")
+            gdcm_image = gdcm_image_reader.GetImage()
+            
+            # determine the correct numpy datatype
+            gdcm_numpy_typemap = {
+                gdcm.PixelFormat.INT8:     numpy.int8,
+                gdcm.PixelFormat.UINT8:    numpy.uint8,
+                gdcm.PixelFormat.UINT16:   numpy.uint16,
+                gdcm.PixelFormat.INT16:    numpy.int16,
+                gdcm.PixelFormat.UINT32:   numpy.uint32,
+                gdcm.PixelFormat.INT32:    numpy.int32,
+                gdcm.PixelFormat.FLOAT32:  numpy.float32,
+                gdcm.PixelFormat.FLOAT64:  numpy.float64
+            }
+            gdcm_pixel_format = gdcm_image.GetPixelFormat().GetScalarType()
+            if gdcm_pixel_format in gdcm_numpy_typemap:
+                numpy_dtype = gdcm_numpy_typemap[gdcm_pixel_format]
+            else:
+                raise TypeError('{} is not a GDCM supported pixel format'.format(gdcm_pixel_format))
+            
+            # get the raw data buffer (decompressed, decoded)
+            gdcm_raw_data_buffer = gdcm_image.GetBuffer()
+            
+            # ideally this will be factored out 
+            # in the case of a lookup table try to apply it
+            # if gdcm_image.GetPhotometricInterpretation().GetType() == gdcm.PhotometricInterpretation.PALETTE_COLOR:
+            #     gdcm_buf_len = gdcm_image.GetBufferLength()
+            #     # apply lookup table if possible
+            #     gdcm_lutfilt = gdcm.ImageApplyLookupTable();
+            #     gdcm_lutfilt.SetInput(gdcm_image);
+            #     if not gdcm_lutfilt.Apply():
+            #         warnings.warn("GDCM failed to apply LUT")
+            #     gdcm_lutfiltered_pixmap = gdcm_lutfilt.GetOutputAsPixmap()
+            #     gdcm_raw_data_buffer = gdcm_lutfiltered_pixmap.GetBuffer()
+            #     
+            #     # amount of samples per pixel can change
+            #     self.SamplesPerPixel = gdcm_lutfiltered_pixmap.GetBufferLength() / gdcm_buf_len
+            #     self.PlanarConfiguration = 0
+            
+            # if GDCM indicates that a byte swap is in order, make sure to inform numpy as well
+            if gdcm_image.GetNeedByteSwap():
+                numpy_dtype.newbyteorder('S')
+
+            # FIXME 
+            # GDCM returns char* as type str. In the case of Python 3 this needs to be encoded back into a byte array.
+            # This is no problem in Python 3 as strings are byte strings by default there.
+            # if we use self.PixelData directly as the FIXME above indicates this will no longer be an issue
+            if sys.version_info >= (3, 0):
+                py3_encoding_succeeded = False
+                for encoding in ["utf-8", "mbcs", sys.getfilesystemencoding()]:
```

Comment:
It took me a while to realize that you are correct. I'll update that next Monday.